src/main/java/net/minecraftforge/event/entity/player/PlayerEvent.java

```diff
@@ -462,6 +463,50 @@ public ItemStack getSmelting()
         }
     }
 
+    /**
+     *
+     * Event fired before the player is accepted to the world
+     *
+     */
+    @Cancelable
+    public static class PrePlayerLoginEvent extends PlayerEvent {
+        private Component denyMessage = null;
+        private final Connection connection;
+
+        public PrePlayerLoginEvent(Player player, Connection connection)
+        {
+            super(player);
+
+            this.connection = connection;
+        }
+
+        @Override
+        public void setCanceled(boolean cancel)
+        {
+            if (this.isCanceled() && !cancel)
+            {
+                throw new RuntimeException("You cannot un-cancel the PrePlayerLogin event");
+            }
+
+            super.setCanceled(cancel);
+        }
+
+        public Connection getConnection()
+        {
+            return this.connection;
+        }
+
+        public Component getDenyMessage()
+        {
+            return this.denyMessage;
+        }
+
+        public void setDenyMessage(Component denyMessage)
```

Comment:
This should have documentation.

Reply:
Added documentation, and also noticed missing default value. So I added a default value and then explained in the documentation that it only exists to prevent future NPEs

Reply:
If a NPE is possible for null values, then setDenyMessage should throw when called with null

Reply:
Added an `Objects#requireNonNull` check before assigning variable (I believe this should throw an error if null)